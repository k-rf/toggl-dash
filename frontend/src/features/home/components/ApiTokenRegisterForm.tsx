import { InputAdornment, Stack } from "@mui/material";
import { KeyboardEventHandler, MouseEventHandler, useState } from "react";

import { Form, HandleSubmit, Methods } from "~/components/Compositions";
import { ElButton, ElTextField } from "~/components/Elements";
import { ElIconButton } from "~/components/Elements/ElIconButton";
import { ElTypography } from "~/components/Elements/ElTypography";
import { VisibilityIcon, VisibilityOffIcon } from "~/components/Icons";
import { z } from "~/lib/zod";

const schema = z.object({
  apiToken: z.string().min(1),
});

type ApiTokenValues = z.infer<typeof schema>;

type Props = { onSubmit?: (value: ApiTokenValues) => Promise<void> | void };

export const ApiTokenRegisterForm = (props: Props) => {
  const [visible, setVisible] = useState(false);

  const submit = (methods: Methods<ApiTokenValues>) => {
    return methods.handleSubmit(async (value) => {
      methods.reset({ apiToken: "" });

      await props.onSubmit?.(value);
    });
  };

  const handleSubmit: HandleSubmit<ApiTokenValues, MouseEventHandler> = (methods) => {
    return submit(methods);
  };

  const handleKeyDown: HandleSubmit<ApiTokenValues, KeyboardEventHandler> = (methods) => {
    return (e) => {
      if (e.code === "Enter") {
        e.preventDefault();

        if (e.ctrlKey) {
          return submit(methods)(e);
        }
      }
    };
  };

  const handleShow = () => {
    setVisible(true);
  };

  const handleMask = () => {
    setVisible(false);
  };

  return (
    <Form<ApiTokenValues> options={{ mode: "all" }} schema={schema}>
      {(methods) => {
        const { error } = methods.getFieldState("apiToken");
        const {
          errors: { apiToken },
          isValid,
        } = methods.formState;

        return (
          <Stack spacing={1}>
            <ElTypography>Toggl の API トークンを入力してください</ElTypography>
            <ElTextField
              fullWidth
              placeholder="API トークン"
              type={visible ? "text" : "password"}
              error={Boolean(error)}
              onKeyDown={handleKeyDown(methods)}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ElIconButton
                      disableTouchRipple
                      onMouseDown={handleShow}
                      onMouseUp={handleMask}
                      onMouseLeave={handleMask}
                      data-testid="api-token-register-form-visibility-toggle-icon-button"
                    >
                      {visible ? <VisibilityOffIcon /> : <VisibilityIcon />}
                    </ElIconButton>
                  </InputAdornment>
                ),
              }}
              registration={methods.register("apiToken")}
            />
            <ElButton
              disabled={!isValid}
              onClick={handleSubmit(methods)}
              data-testid="api-token-register-form-submit-button"
            >
              保存する
            </ElButton>
            <ElTypography variant="body2" color="error">
              {apiToken?.message}
            </ElTypography>
          </Stack>
        );
      }}
    </Form>
  );
};

export type { Props as ApiTokenRegisterFormProps };
