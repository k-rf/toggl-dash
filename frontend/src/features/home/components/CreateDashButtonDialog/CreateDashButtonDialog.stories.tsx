import type { Meta, StoryObj } from "@storybook/react";
// import { userEvent, waitFor, within } from "@storybook/testing-library";

import { CreateDashButtonDialog } from "./CreateDashButtonDialog";

export default { component: CreateDashButtonDialog, args: { open: true } } as Meta<
  typeof CreateDashButtonDialog
>;

type Story = StoryObj<typeof CreateDashButtonDialog>;

export const Default: Story = {};

// Play Utils
// -----------------------------------------------------------------------------

// TODO:
// type Context = {};

// function playUtils(play: (context: Context) => Promise<void> | void) {
//   return async ({ canvasElement }: { canvasElement: HTMLElement }) => {};
// }
