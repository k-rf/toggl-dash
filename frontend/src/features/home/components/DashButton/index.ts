export * from "./DashButton";
export * from "./DashButtonDetailListItem";
