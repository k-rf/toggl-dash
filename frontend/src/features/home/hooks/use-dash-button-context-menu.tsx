import React, { useCallback, useState } from "react";

import { ContextMenu } from "../components/ContextMenu/ContextMenu";

export const useDashButtonContextMenu = (dashButtonId: string) => {
  const [position, setPosition] = useState<{ mouseX: number; mouseY: number } | null>(null);

  const handleOpen = useCallback((e: React.MouseEvent) => {
    e.preventDefault();

    setPosition({ mouseX: e.clientX, mouseY: e.clientY });
  }, []);

  const handleClose = useCallback(() => {
    setPosition(null);
  }, []);

  return {
    handleOpen,
    render: () => (
      <ContextMenu
        open={Boolean(position)}
        onClose={handleClose}
        onContextMenu={handleOpen}
        anchorPosition={position ? { top: position.mouseY, left: position.mouseX } : undefined}
        dashButtonId={dashButtonId}
      />
    ),
  };
};
