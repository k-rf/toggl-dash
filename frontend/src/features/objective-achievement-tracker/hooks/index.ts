export * from "./use-available-time-form";
export * from "./use-available-time-form-dialog";
export * from "./use-objective-form";
