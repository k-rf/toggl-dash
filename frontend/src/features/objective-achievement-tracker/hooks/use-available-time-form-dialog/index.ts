export * from "./use-available-time-form-dialog";
