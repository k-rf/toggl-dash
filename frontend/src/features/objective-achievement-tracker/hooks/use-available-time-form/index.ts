export * from "./available-time-form-schema";
export * from "./use-available-time-form";
