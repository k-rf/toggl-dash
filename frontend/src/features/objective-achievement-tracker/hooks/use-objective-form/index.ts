export * from "./objective-form-schema";
export * from "./use-objective-form";
