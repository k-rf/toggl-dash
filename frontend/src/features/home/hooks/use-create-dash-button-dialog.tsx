import { useCallback, useState } from "react";

import { CreateDashButtonDialog } from "../components/CreateDashButtonDialog";

export const useCreateDashButtonDialog = () => {
  const [open, setOpen] = useState(false);

  const handleOpen = useCallback(() => {
    setOpen(true);
  }, []);

  const handelClose = useCallback(() => {
    setOpen(false);
  }, []);

  return {
    handleOpen,
    render: () => <CreateDashButtonDialog open={open} onClose={handelClose} />,
  };
};
