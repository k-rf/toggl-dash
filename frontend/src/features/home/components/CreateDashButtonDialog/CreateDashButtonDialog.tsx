import { zodResolver } from "@hookform/resolvers/zod";
import { Stack } from "@mui/material";
import { useState } from "react";
import { useForm } from "react-hook-form";
import { z } from "zod";

import { RhfTextField } from "~/components/Compositions";
import { ElButton } from "~/components/Elements";
import {
  ElDialog,
  ElDialogActions,
  ElDialogContent,
  ElDialogProps,
  ElDialogTitle,
} from "~/components/Elements/ElDialog";
import { ElMenuItem } from "~/components/Elements/ElMenu";
import { useGetTogglClientsLazyQuery, useGetTogglProjectsByClientLazyQuery } from "~/graphql";
import { useLoading } from "~/hooks/use-loading";

const schema = z.object({
  clientId: z.preprocess(Number, z.number().int().min(1)).transform(String),
  projectId: z.preprocess(Number, z.number().int().min(1)).transform(String),
  description: z.string().min(1),
});

type Props = Pick<ElDialogProps, "open" | "onClose">;

export const CreateDashButtonDialog = (props: Props) => {
  const [getTogglClients, { data: togglClients }] = useGetTogglClientsLazyQuery();
  const [getTogglProjectsByClient, { data: togglProjects }] =
    useGetTogglProjectsByClientLazyQuery();

  const { handleSubmit, reset, formState, control, getValues } = useForm<z.infer<typeof schema>>({
    resolver: zodResolver(schema),
    mode: "all",
  });

  const { loadingState } = useLoading(["client", "project"]);

  const [togglClientSelectOpen, setTogglClientSelectOpen] = useState(false);
  const [togglProjectSelectOpen, setTogglProjectOpen] = useState(false);

  return (
    <ElDialog open={props.open} onClose={props.onClose}>
      <ElDialogTitle>ダッシュボタンを作成する</ElDialogTitle>

      <ElDialogContent sx={{ overflow: "visible" }}>
        <Stack spacing={2}>
          <RhfTextField
            type="select"
            control={control}
            name="clientId"
            defaultValue=""
            label="クライアント"
            loading={loadingState("client").get()}
            open={togglClientSelectOpen}
            onOpen={async () => {
              await loadingState("client").handle(async () => getTogglClients());

              setTogglClientSelectOpen(true);
            }}
            onClick={() => setTogglClientSelectOpen(false)}
          >
            {togglClients?.togglClientAll.map((client) => (
              <ElMenuItem key={client.id} value={client.id}>
                {client.name}
              </ElMenuItem>
            ))}
          </RhfTextField>

          <RhfTextField
            type="select"
            control={control}
            name="projectId"
            defaultValue=""
            label="プロジェクト"
            disabled={!getValues("clientId")}
            loading={loadingState("project").get()}
            open={togglProjectSelectOpen}
            onOpen={async () => {
              await loadingState("project").handle(async () =>
                getTogglProjectsByClient({ variables: { id: Number(getValues("clientId")) } })
              );

              setTogglProjectOpen(true);
            }}
            onClick={() => setTogglProjectOpen(false)}
          >
            {togglProjects?.togglProjectByClient.map((project) => (
              <ElMenuItem key={project.id} value={project.id}>
                {project.name}
              </ElMenuItem>
            ))}
          </RhfTextField>

          <RhfTextField
            type="text"
            control={control}
            name="description"
            defaultValue=""
            label="詳細"
            error={!!formState.errors.description}
          />
        </Stack>
      </ElDialogContent>

      <ElDialogActions>
        <ElButton
          variant="text"
          color="inherit"
          onClick={(e) => {
            reset();
            props.onClose?.(e, "backdropClick");
          }}
        >
          キャンセル
        </ElButton>
        <ElButton
          disabled={!formState.isValid}
          onClick={handleSubmit(
            (data) => {
              reset();
              console.log(data);
            },
            (errors) => {
              console.log(errors);
            }
          )}
        >
          作成する
        </ElButton>
      </ElDialogActions>
    </ElDialog>
  );
};
