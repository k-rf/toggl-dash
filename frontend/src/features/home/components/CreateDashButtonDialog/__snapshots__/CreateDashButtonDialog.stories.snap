// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/Home/Components/CreateDashButtonDialog/CreateDashButtonDialog > Default 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r3:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-u5nrt4-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r3:"
        >
          ダッシュボタンを作成する
        </h2>
        <div
          class="MuiDialogContent-root css-1cqcnnz-MuiDialogContent-root"
        >
          <div
            class="css-1p5q5e5-MuiStack-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-lvpc63-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="clientId"
                id="clientId-label"
              >
                クライアント
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="clientId-label clientId"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-ecdluo-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="clientId"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="clientId"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      クライアント
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-lvpc63-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="projectId"
                id="projectId-label"
              >
                プロジェクト
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  aria-disabled="true"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="projectId-label projectId"
                  class="MuiSelect-select MuiSelect-outlined Mui-disabled MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-ecdluo-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="projectId"
                  role="button"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  disabled=""
                  name="projectId"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined Mui-disabled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      プロジェクト
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-lvpc63-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="description"
                id="description-label"
              >
                詳細
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1fa36cw-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="description"
                  name="description"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      詳細
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-pzz4qf-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            キャンセル
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-a3cr8t-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            作成する
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
